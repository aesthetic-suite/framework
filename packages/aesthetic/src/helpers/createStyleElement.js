/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

export default function createStyleElement(id: string): HTMLElement {
  const style = document.createElement('style');
  style.type = 'text/css';
  style.id = `aesthetic-${id}`;

  // $FlowIgnore We know the document exists
  document.head.appendChild(style);

  return style;
}
