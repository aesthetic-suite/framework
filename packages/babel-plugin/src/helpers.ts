import { types as t } from '@babel/core';
import { createDebugger } from '@boost/debug';

export const debug = createDebugger('aesthetic:babel');

export function isFunction(node: t.Node) {
  return t.isFunctionExpression(node) || t.isArrowFunctionExpression(node);
}
