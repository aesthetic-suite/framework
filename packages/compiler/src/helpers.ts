import { ColorShade } from '@aesthetic/system';

export function formatShade(value: number | string): ColorShade {
	return String(value).padStart(2, '0').slice(0, 2) as ColorShade;
}

export function formatUnit(value: number): string {
	return value.toFixed(2).replace('.00', '');
}
