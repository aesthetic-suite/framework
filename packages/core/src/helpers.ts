import { Theme } from '@aesthetic/system';
import { objectLoop } from '@aesthetic/utils';
import { SheetParams } from './types';

export function createCacheKey(params: SheetParams, type: string): string {
  let key = type;

  // Since all other values are scalars, we can just join the values.
  // This is 3x faster than JSON.stringify(), and 1.5x faster than Object.values()!
  objectLoop(params, (value) => {
    key += value;
  });

  return key;
}

export function createDefaultParams(theme: Theme<{}>, params: SheetParams): Required<SheetParams> {
  return {
    contrast: theme.contrast,
    deterministic: false,
    direction: 'ltr',
    scheme: theme.scheme,
    theme: theme.name,
    unit: 'px',
    vendor: false,
    ...params,
  };
}
