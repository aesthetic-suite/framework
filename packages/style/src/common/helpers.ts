/* eslint-disable no-magic-numbers */

import { Value, MaybeValue, CSS, Sheet } from '@aesthetic/types';
import { hyphenate } from '@aesthetic/utils';
import { IMPORT_RULE, STYLE_RULE } from './constants';

export function insertRule(sheet: Sheet, rule: CSS, index?: number): number {
  try {
    return sheet.insertRule(rule, index ?? sheet.cssRules.length);
  } catch {
    // Vendor prefixed properties, pseudos, etc, that are inserted
    // into different vendors will trigger a failure. For example,
    // `-moz` or `-ms` being inserted into WebKit.
    // There's no easy way around this, so let's just ignore the
    // error so that subsequent styles are inserted.
    // istanbul ignore next
    return -1;
  }
}

export function insertAtRule(sheet: Sheet, rule: CSS): number {
  const { length } = sheet.cssRules;
  let index = 0;

  // At-rules must be inserted before normal style rules.
  for (let i = 0; i <= length; i += 1) {
    index = i;

    if (sheet.cssRules[i]?.type === STYLE_RULE) {
      break;
    }
  }

  return insertRule(sheet, rule, index);
}

export function insertImportRule(sheet: Sheet, rule: CSS): number {
  const { length } = sheet.cssRules;
  let index = 0;

  // Import rules must be inserted at the top of the style sheet,
  // but we also want to persist the existing order.
  for (let i = 0; i <= length; i += 1) {
    index = i;

    if (sheet.cssRules[i]?.type !== IMPORT_RULE) {
      break;
    }
  }

  return insertRule(sheet, rule, index);
}

export function isAtRule(value: string): boolean {
  return value[0] === '@';
}

export function isImportRule(value: string): boolean {
  return value.slice(0, 7) === '@import';
}

export function isMediaRule(value: string): boolean {
  return value.slice(0, 6) === '@media';
}

export function isNestedSelector(value: string): boolean {
  const char = value[0];

  return (
    char === ':' ||
    char === '[' ||
    char === '>' ||
    char === '~' ||
    char === '+' ||
    char === '*' ||
    char === '|'
  );
}

export function isSupportsRule(value: string): boolean {
  return value.slice(0, 9) === '@supports';
}

const unitlessProperties = new Set<string>();

[
  'animationIterationCount',
  'borderImage',
  'borderImageOutset',
  'borderImageSlice',
  'borderImageWidth',
  'columnCount',
  'columns',
  'flex',
  'flexGrow',
  'flexPositive',
  'flexShrink',
  'flexNegative',
  'flexOrder',
  'fontWeight',
  'gridArea',
  'gridRow',
  'gridRowEnd',
  'gridRowSpan',
  'gridRowStart',
  'gridColumn',
  'gridColumnEnd',
  'gridColumnSpan',
  'gridColumnStart',
  'lineClamp',
  'lineHeight',
  'maskBorder',
  'maskBorderOutset',
  'maskBorderSlice',
  'maskBorderWidth',
  'opacity',
  'order',
  'orphans',
  'tabSize',
  'widows',
  'zIndex',
  'zoom',
  // SVG
  'fillOpacity',
  'floodOpacity',
  'stopOpacity',
  'strokeDasharray',
  'strokeDashoffset',
  'strokeMiterlimit',
  'strokeOpacity',
  'strokeWidth',
].forEach((property) => {
  unitlessProperties.add(property);
  unitlessProperties.add(hyphenate(property));
});

export function isUnitlessProperty(property: string): boolean {
  return unitlessProperties.has(property);
}

export function isValidValue(property: string, value: MaybeValue | boolean): value is Value {
  if (value === null || value === undefined || value === true || value === false || value === '') {
    if (__DEV__) {
      // eslint-disable-next-line no-console
      console.warn(`Invalid value "${value}" for "${property}".`);
    }

    return false;
  }

  return true;
}

export function isVariable(value: string): boolean {
  return value.slice(0, 2) === '--';
}

export function joinQueries(prev: string | undefined, next: string): string {
  if (prev) {
    return `${prev} and ${next}`;
  }

  return next;
}
