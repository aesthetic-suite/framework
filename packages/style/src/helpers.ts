/* eslint-disable no-magic-numbers */

import { Value, MaybeValue } from '@aesthetic/types';
import { hyphenate } from '@aesthetic/utils';

export function isAtRule(value: string): boolean {
  return value[0] === '@';
}

export function isImportRule(value: string): boolean {
  return value.slice(0, 7) === '@import' && value[7] === ' ';
}

export function isNestedSelector(value: string): boolean {
  const char = value[0];

  return (
    char === ':' ||
    char === '[' ||
    char === '>' ||
    char === '~' ||
    char === '+' ||
    char === '*' ||
    char === '|'
  );
}

const unitlessProperties = new Set<string>();

[
  'animationIterationCount',
  'borderImage',
  'borderImageOutset',
  'borderImageSlice',
  'borderImageWidth',
  'columnCount',
  'columns',
  'flex',
  'flexGrow',
  'flexPositive',
  'flexShrink',
  'flexNegative',
  'flexOrder',
  'fontWeight',
  'gridArea',
  'gridRow',
  'gridRowEnd',
  'gridRowSpan',
  'gridRowStart',
  'gridColumn',
  'gridColumnEnd',
  'gridColumnSpan',
  'gridColumnStart',
  'lineClamp',
  'lineHeight',
  'maskBorder',
  'maskBorderOutset',
  'maskBorderSlice',
  'maskBorderWidth',
  'opacity',
  'order',
  'orphans',
  'tabSize',
  'widows',
  'zIndex',
  'zoom',
  // SVG
  'fillOpacity',
  'floodOpacity',
  'stopOpacity',
  'strokeDasharray',
  'strokeDashoffset',
  'strokeMiterlimit',
  'strokeOpacity',
  'strokeWidth',
].forEach((property) => {
  unitlessProperties.add(property);
  unitlessProperties.add(hyphenate(property));
});

export function isUnitlessProperty(property: string): boolean {
  return unitlessProperties.has(property);
}

export function isValidValue(property: string, value: MaybeValue | boolean): value is Value {
  if (value === null || value === undefined || value === true || value === false || value === '') {
    if (__DEV__) {
      console.warn(`Invalid value "${value}" for "${property}".`);
    }

    return false;
  }

  return true;
}

export function isVariable(value: string): boolean {
  return value.slice(0, 2) === '--';
}
