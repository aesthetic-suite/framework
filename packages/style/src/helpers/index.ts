import createAtomicCacheKey from './createAtomicCacheKey';
import formatConditions from './formatConditions';
import formatDeclaration from './formatDeclaration';
import formatDeclarationBlock from './formatDeclarationBlock';
import formatRule from './formatRule';
import formatVariable from './formatVariable';
import formatVariableName from './formatVariableName';
import getDocumentStyleSheet from './getDocumentStyleSheet';
import isAtRule from './isAtRule';
import isFontFaceFule from './isFontFaceFule';
import isImportRule from './isImportRule';
import isInvalidValue from './isInvalidValue';
import isKeyframesRule from './isKeyframesRule';
import isMediaRule from './isMediaRule';
import isNestedSelector from './isNestedSelector';
import isSupportsRule from './isSupportsRule';
import isUnitlessProperty from './isUnitlessProperty';
import isVariable from './isVariable';

export {
  createAtomicCacheKey,
  formatConditions,
  formatDeclaration,
  formatDeclarationBlock,
  formatRule,
  formatVariable,
  formatVariableName,
  getDocumentStyleSheet,
  isAtRule,
  isFontFaceFule,
  isImportRule,
  isInvalidValue,
  isKeyframesRule,
  isMediaRule,
  isNestedSelector,
  isSupportsRule,
  isUnitlessProperty,
  isVariable,
};
