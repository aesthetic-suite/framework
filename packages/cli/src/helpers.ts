import { CONFIG_FOLDER } from '@aesthetic/compiler';
import { Path } from '@boost/common';
import { CWD } from './constants';

export function getConfigFolderDir(name: string): Path {
  return new Path(CWD, CONFIG_FOLDER, name);
}

export function validateSystemName(value: string) {
  if (value && !value.match(/^[-a-z0-9]+$/giu)) {
    throw new Error('Name may only contain numbers, letters, and dashes.');
  }
}
