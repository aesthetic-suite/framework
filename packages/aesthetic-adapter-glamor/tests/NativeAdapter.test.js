/* eslint-disable sort-keys */

import { css, speedy, flush } from 'glamor';
import GlamorAdapter from '../src/NativeAdapter';
import {
  FONT_ROBOTO_FLAT_SRC,
  FONT_CIRCULAR_MULTIPLE_FLAT_SRC,
  KEYFRAME_FADE,
  SYNTAX_NATIVE_PARTIAL,
  SYNTAX_ATTRIBUTE,
  SYNTAX_PSEUDO,
} from '../../../tests/mocks';
import { renderGlamorStyles } from '../../../tests/helpers';

describe('aesthetic-adapter-glamor/NativeAdapter', () => {
  let instance;

  beforeEach(() => {
    flush();
    speedy(true);

    instance = new GlamorAdapter();
  });

  it('transforms style declarations into class names', () => {
    expect(instance.transform(instance.create(SYNTAX_NATIVE_PARTIAL).button)).toBe('css-1n8n9n3');
  });

  it('can transform dynamic styles', () => {
    expect(
      instance.transform({
        width: 10,
        height: 10,
      }),
    ).toBe('css-s3feo0');

    expect(renderGlamorStyles(instance)).toMatchSnapshot();
  });

  it('combines different style declarations into unique class names', () => {
    const sheet = instance.create({
      foo: {
        color: 'red',
        display: 'block',
      },
      bar: {
        color: 'green',
        margin: 5,
      },
      baz: {
        color: 'blue',
        padding: 5,
      },
    });

    expect(instance.transform(sheet.foo)).toBe('css-1r6w0zj');
    expect(instance.transform(sheet.bar)).toBe('css-otvq3e');
    expect(instance.transform(sheet.baz)).toBe('css-1s3i2b7');
    expect(instance.transform(sheet.foo, sheet.baz)).toBe('css-1pp6oml');
    expect(instance.transform(sheet.bar, sheet.foo)).toBe('css-h4vsx0');
  });

  it('handles attribute selectors', () => {
    expect(instance.transform(instance.create(SYNTAX_ATTRIBUTE).attr)).toBe('css-blwook');

    expect(renderGlamorStyles(instance)).toMatchSnapshot();
  });

  it('handles descendant selectors', () => {
    expect(
      instance.transform(
        instance.create({
          list: {
            margin: 0,
            padding: 0,
            '& > li': {
              listStyle: 'bullet',
            },
          },
        }).list,
      ),
    ).toBe('css-191gc4j');

    expect(renderGlamorStyles(instance)).toMatchSnapshot();
  });

  it('handles pseudo selectors', () => {
    expect(instance.transform(instance.create(SYNTAX_PSEUDO).pseudo)).toBe('css-1g7aevf');

    expect(renderGlamorStyles(instance)).toMatchSnapshot();
  });

  it('handles fallbacks', () => {
    expect(
      instance.transform(
        instance.create({
          fallback: {
            background: ['red', 'linear-gradient(...)'],
            display: ['box', 'flex-box', 'flex'],
          },
        }).fallback,
      ),
    ).toBe('css-1nxkcks');

    // Verified it ran but fallbacks don't appear in the output
    expect(renderGlamorStyles(instance)).toMatchSnapshot();
  });

  it('handles font faces', () => {
    css.fontFace(FONT_ROBOTO_FLAT_SRC);

    expect(
      instance.transform(
        instance.create({
          font: {
            fontFamily: 'Roboto',
            fontSize: 20,
          },
        }).font,
      ),
    ).toBe('css-1x6s9dk');

    expect(renderGlamorStyles(instance)).toMatchSnapshot();
  });

  it('handles multiple font faces', () => {
    FONT_CIRCULAR_MULTIPLE_FLAT_SRC.forEach(font => {
      css.fontFace(font);
    });

    expect(
      instance.transform(
        instance.create({
          font: {
            fontFamily: 'Circular',
            fontSize: 20,
          },
        }).font,
      ),
    ).toBe('css-ydlv7p');

    expect(renderGlamorStyles(instance)).toMatchSnapshot();
  });

  it('handles animations', () => {
    expect(
      instance.transform(
        instance.create({
          animation: {
            animationName: css.keyframes('fade', KEYFRAME_FADE),
            animationDuration: '3s, 1200ms',
            animationIterationCount: 'infinite',
          },
        }).animation,
      ),
    ).toBe('css-s8bawe');

    expect(renderGlamorStyles(instance)).toMatchSnapshot();
  });

  it('handles media queries', () => {
    expect(
      instance.transform(
        instance.create({
          media: {
            color: 'red',
            '@media (min-width: 300px)': {
              color: 'blue',
            },
            '@media (max-width: 1000px)': {
              color: 'green',
            },
          },
        }).media,
      ),
    ).toBe('css-rr71yy');

    expect(renderGlamorStyles(instance)).toMatchSnapshot();
  });

  it('handles supports', () => {
    expect(
      instance.transform(
        instance.create({
          sup: {
            display: 'block',
            '@supports (display: flex)': {
              display: 'flex',
            },
            '@supports not (display: flex)': {
              float: 'left',
            },
          },
        }).sup,
      ),
    ).toBe('css-1sp1mbh');

    // Verified it ran but supports don't appear in the output
    expect(renderGlamorStyles(instance)).toMatchSnapshot();
  });
});
