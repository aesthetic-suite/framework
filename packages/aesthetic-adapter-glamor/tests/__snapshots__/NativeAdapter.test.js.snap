// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aesthetic-adapter-glamor/NativeAdapter can transform dynamic styles 1`] = `".css-s3feo0,[data-css-s3feo0] {width: 10px; height: 10px;}"`;

exports[`aesthetic-adapter-glamor/NativeAdapter handles animations 1`] = `"@-webkit-keyframes fade_1q3syk4 {from {opacity: 0;}to {opacity: 1;} }@-moz-keyframes fade_1q3syk4 {from {opacity: 0;}to {opacity: 1;} }@-o-keyframes fade_1q3syk4 {from {opacity: 0;}to {opacity: 1;} }@keyframes fade_1q3syk4 {from {opacity: 0;}to {opacity: 1;} }.css-s8bawe,[data-css-s8bawe] {animation-name: fade_1q3syk4; animation-duration: 3s, 1200ms; animation-iteration-count: infinite; -webkit-animation-name: fade_1q3syk4; -webkit-animation-duration: 3s, 1200ms; -webkit-animation-iteration-count: infinite;}"`;

exports[`aesthetic-adapter-glamor/NativeAdapter handles attribute selectors 1`] = `".css-blwook,[data-css-blwook] {display: block;}.css-blwook[disabled],[data-css-blwook][disabled] {opacity: 0.5;}"`;

exports[`aesthetic-adapter-glamor/NativeAdapter handles descendant selectors 1`] = `".css-191gc4j,[data-css-191gc4j] {margin: 0; padding: 0;}.css-191gc4j > li,[data-css-191gc4j] > li {list-style: bullet;}"`;

exports[`aesthetic-adapter-glamor/NativeAdapter handles fallbacks 1`] = `".css-1nxkcks,[data-css-1nxkcks] {background: linear-gradient(...); display: flex;}"`;

exports[`aesthetic-adapter-glamor/NativeAdapter handles font faces 1`] = `"@font-face {font-family: Roboto; font-style: normal; font-weight: normal; src: local('Robo'), url('fonts/Roboto.woff2') format('woff2'), url('fonts/Roboto.ttf') format('truetype');}.css-1x6s9dk,[data-css-1x6s9dk] {font-family: Roboto; font-size: 20px;}"`;

exports[`aesthetic-adapter-glamor/NativeAdapter handles media queries 1`] = `".css-rr71yy,[data-css-rr71yy] {color: red;}@media (min-width: 300px) {.css-rr71yy,[data-css-rr71yy] {color: blue;}}@media (max-width: 1000px) {.css-rr71yy,[data-css-rr71yy] {color: green;}}"`;

exports[`aesthetic-adapter-glamor/NativeAdapter handles multiple font faces 1`] = `"@font-face {font-family: Circular; font-style: normal; font-weight: normal; src: url('fonts/Circular.woff2') format('woff2');}@font-face {font-family: Circular; font-style: italic; font-weight: normal; src: url('fonts/Circular-Italic.woff2') format('woff2');}@font-face {font-family: Circular; font-style: normal; font-weight: 300; src: url('fonts/Circular-Light.woff2') format('woff2');}@font-face {font-family: Circular; font-style: normal; font-weight: 700; src: url('fonts/Circular-Bold.woff2') format('woff2');}.css-ydlv7p,[data-css-ydlv7p] {font-family: Circular; font-size: 20px;}"`;

exports[`aesthetic-adapter-glamor/NativeAdapter handles pseudo selectors 1`] = `".css-1g7aevf,[data-css-1g7aevf] {position: fixed;}.css-1g7aevf:hover,[data-css-1g7aevf]:hover {position: static;}.css-1g7aevf::before,[data-css-1g7aevf]::before {position: absolute;}"`;

exports[`aesthetic-adapter-glamor/NativeAdapter handles supports 1`] = `".css-1sp1mbh,[data-css-1sp1mbh] {display: block;}"`;
